Optionally include documentation in GDExtension API dump

rburing opened the pull request:
Language bindings implemented in GDExtension that generate glue code (in the future, C#?) may want to include documentation (for classes, properties, methods, signals, etc.) to improve the user experience in code editors and REPLs. 

This PR adds an option to include that documentation in the `extension_api.json` file.

This is much more convenient than having to obtain and parse ~829 XML files from the Godot source code.

(Ideally the build system for a GDExtension should depend just on having the correct API file, and not on having to run Godot or having the Godot source code.)

* Size of `extension_api.json` generated using `--dump-extension-api`: 5559324 bytes.
* Size of `extension_api.json` generated using `--dump-extension-api-with-docs`: 9121481 bytes.
* The size increase is 65%.

Feedback welcome.

There are quite some loops that could be hidden away by adding new core methods to `ClassDoc`, if desired.

rburing committed:
Optionally include documentation in GDExtension API dump

Bromeon commented on core/extension/extension_api_dump.cpp:
```diff
@@ -479,6 +499,14 @@ Dictionary GDExtensionAPIDump::generate_extension_api() {
 				d["name"] = name;
 				d["value"] = value;
 				d["is_bitfield"] = bitfield;
+				if (p_include_docs) {
+					for (const DocData::ConstantDoc &constant_doc : global_scope_doc->constants) {
+						if (constant_doc.name == name) {
+							d["documentation"] = fix_doc_description(constant_doc.description);
+							break;
+						}
+					}
+				}
```
This runs in O(n<sup>2</sup>) on the number of constants... which I guess is OK? This number is typically low (dozens)?

(Same for other cases)

dsnopek replied:
Personally, I think this is probably fine. This is a "batch process" that's running "offline", so it doesn't need to be fast (within reason). I timed running `--dump-extension-api-with-docs` couple of times locally, and it's taking about 3 seconds on average on my system.

Bromeon reviewed (commented):
As a heavy user of `extension_api.json`, I would welcome this addition. It would mean one less data source and one less format (XML) that needs to be understood.

The XML files also contain some meta-information that is exclusive to them, for example whether a class is experimental. Do you think it makes sense to add this to the JSON as well -- maybe even the base (non-doc) version?

dsnopek commented on main/main.cpp:
```diff
@@ -513,6 +514,7 @@ void Main::print_help(const char *p_binary) {
 	OS::get_singleton()->print("  --build-solutions                 Build the scripting solutions (e.g. for C# projects). Implies --editor and requires a valid project to edit.\n");
 	OS::get_singleton()->print("  --dump-gdextension-interface      Generate GDExtension header file 'gdextension_interface.h' in the current folder. This file is the base file required to implement a GDExtension.\n");
 	OS::get_singleton()->print("  --dump-extension-api              Generate JSON dump of the Godot API for GDExtension bindings named 'extension_api.json' in the current folder.\n");
+	OS::get_singleton()->print("  --dump-extension-api-with-docs    Generate JSON dump of the Godot API (including documentation) for GDExtension bindings named 'extension_api.json' in the current folder.\n");
```
This feels a little too repetitive: when skimming quickly it's hard to spot the difference between this one and the one above. Many of the existing option descriptions make reference to other options - maybe this option could refer to the one above? For example, "Generate JSON dump like the previous option, but including documentation."

dsnopek commented on main/main.cpp:
```diff
@@ -513,6 +514,7 @@ void Main::print_help(const char *p_binary) {
 	OS::get_singleton()->print("  --dump-gdextension-interface      Generate GDExtension header file 'gdextension_interface.h' in the current folder. This file is the base file required to implement a GDExtension.\n");
 	OS::get_singleton()->print("  --dump-extension-api              Generate JSON dump of the Godot API for GDExtension bindings named 'extension_api.json' in the current folder.\n");
+	OS::get_singleton()->print("  --dump-extension-api-with-docs    Generate JSON dump of the Godot API (including documentation) for GDExtension bindings named 'extension_api.json' in the current folder.\n");
 	OS::get_singleton()->print("  --validate-extension-api <path>   Validate an extension API file dumped (with the option above) from a previous version of the engine to ensure API compatibility. If incompatibilities or errors are detected, the return code will be non zero.\n");
```
This option is making reference to the `--dump-extension-api` option, so it should probably be updated to refer to both of them. For example:

```suggestion
	OS::get_singleton()->print("  --validate-extension-api <path>   Validate an extension API file dumped (with one of the two previous option) from a previous version of the engine to ensure API compatibility. If incompatibilities or errors are detected, the return code will be non zero.\n");
```

dsnopek reviewed (approved):
Thanks, this is a great idea!

I tested locally, and it seems to be working, although I only skimmed through the JSON, I didn't scrutinize it in depth.

The code looks good to me as well! I only have some minor notes about the command-line help info.

rburing commented:
Thanks for the timely reviews!

@Bromeon 

> The XML files also contain some meta-information that is exclusive to them, for example whether a class is experimental. Do you think it makes sense to add this to the JSON as well -- maybe even the base (non-doc) version?

It makes sense to me, but I would prefer not to touch it in this PR.

We can discuss it with the GDExtension team on RocketChat.

akien-mga commented:
Thanks!

maiself commented:
Working with python:

![Screenshot from 2023-09-26 17-12-48](https://github.com/godotengine/godot/assets/6334580/af01b6ea-0803-4854-b9b5-bb26bc636543)

Parsing the bbcode is a little awkward, but not too bad. This is really cool, thanks!

akien-mga commented:
> Parsing the bbcode is a little awkward, but not too bad.

I assume many language bindings will use Python for codegen from the json. Maybe we could have a small python library that can be used for parsing the bbcode in each language binding?

GeorgeS2019 commented:
@rburing 
Thanks for your contribution

We are starting to see a path to [create GDExtension C# Bindings](https://github.com/RhubarbVR/GDExtensionCSharp) using your merge

rburing commented:
For those who found this PR and want to use the feature: note that in https://github.com/godotengine/godot/pull/83318 we changed the key name from `"documentation"` to `"description"` (for internal consistency) and we added `"brief_description"` to classes and built-in classes.

This will allow generating even more detailed documentation.
